### DISCUSSION TITLE
Running on multiple cores

### DISCUSSION THREAD

**Discussion by LumaRay** [matt-42/lithium#117] [Q&A] [answered]
Hey everyone!

I try to experiment with Lithium in Ubuntu 18.04 on a virtual machine.
1. I run Lithium in Docker by `./li run ./test-lithium.cc`
2. I run wrk test: `~/wrk/wrk -t4 -c400 -d30s http://127.0.0.1:7774` - but it gives poor results (max 250MB/sec compared to rust/hyper which gives up to 900MB/sec)

While testing I can see in htop that only 1 of 4 cores is used, while rust/hyper uses all of them.

How can I configure Lithium to run on all cores?

Thanks!

**Comment by matt-42** [comment, marked as answer]
You can use the s::nthreads options:
    `http_serve(api, port, s::nthreads = __number_of_threads_you_want__);`

**Reply by LumaRay** [reply]
Thanks Matt,
I set 4 threads, but unfortunately it still uses 1 core:
![image](https://user-images.githubusercontent.com/2438226/186641467-25de2b7b-87be-4487-bab9-08ba308f5c5c.png)


![image](https://user-images.githubusercontent.com/2438226/186641175-f5f0fe8a-5ef2-4e6d-a4d1-6e697f5b12c2.png)

**Reply by LumaRay** [reply]
Maybe this is the reason?..
![image](https://user-images.githubusercontent.com/2438226/186642486-49f23807-2706-4c08-b241-9337167365f5.png)

**Reply by Mickael-van-der-Beek** [reply]
@LumaRay Have you been able to solve your issue?

We are seeing the same behaviour as you.
A 12 core Debian 11 server barely uses any memory or CPU while benchmarking and show very poor performance results.

The cmake error shouldn't be an issue since a thread library was found: https://stackoverflow.com/a/64515538/18635875

**Reply by LumaRay** [reply]
> @LumaRay Have you been able to solve your issue?

Nope, I thought the author might fix this... I have made some benchmarks, although not very accurate, I guess, but Lithium might beat the others if properly updated. https://github.com/LumaRay/test-simple-web-server

**Reply by Mickael-van-der-Beek** [reply]
@matt-42 Since there are basically no options that need to be tweaked or that are tweaked in your or TechEmpower's benchmark, would you be able to help us out finding out why Lithium wouldn't use the full amount of available resources?

As additional information we set `net.core.somaxconn=2048` and `net.ipv4.tcp_max_syn_backlog=10000`.

**Reply by Mickael-van-der-Beek** [reply]
@matt-42 This ticket is related to the issue I'm mentioning: https://github.com/matt-42/lithium/issues/122
